core/src/blockchain/blockchain.rs

```diff
@@ -212,6 +212,48 @@ impl BlockChain {
         }
     }
 
+    /// Force update the best block as the given block hash.
+    ///
+    /// This will not change the highest block chain. This means it is possible
+    pub fn update_best_block_hash(&self, batch: &mut DBTransaction, block_hash: H256) -> ImportRoute {
+        // FIXME: If it is possible, double check with the consensus engine.
+
+        assert!(self.pending_best_block_hash.read().is_none());
+        let block_detail = self.block_details(&block_hash).expect("The given hash should exist");
+        let parent_hash = block_detail.parent;
+
+        let prev_best_hash = self.best_block_hash();
+        let best_block_changed = if block_hash != prev_best_hash {
+            let route = tree_route(self, prev_best_hash, parent_hash)
+                .expect("blocks being imported always within recent history; qed");
+
+            let best_block =
+                self.block(&block_hash).expect("Best block is already imported as a branch").rlp().as_raw().to_vec();
+
+            match route.retracted.len() {
+                0 => BestBlockChanged::CanonChainAppended {
+                    best_block,
+                },
+                _ => BestBlockChanged::BranchBecomingCanonChain {
+                    tree_route: route,
+                    best_block,
+                },
+            }
+        } else {
+            BestBlockChanged::None
+        };
+
+        self.body_db.update_best_block(batch, &best_block_changed);
```

Comment:
We should also update the block's hash entry(BlockNumber -> BlockHash).
https://github.com/CodeChain-io/codechain/blob/36b57fc83f0563a4c880edf257d2374fe72852b2/core/src/blockchain/headerchain.rs#L154

Reply:
Added `update_best_block_in_tendermint` in HeaderChain.